File: frontend/src/shared/schemas/settings.schemas.ts
/**
 * @fileoverview Settings-related Zod schemas.
 *
 * Defines validation schemas for user settings including API keys
 * and LLM provider configuration.
 *
 * @module schemas/settings
 */

import { z } from 'zod';

// =============================================================================
// API Key Schemas
// =============================================================================

/** AI Provider types */
export const aiProviderSchema = z.enum(['openai', 'anthropic', 'gemini']);
export type AIProvider = z.infer<typeof aiProviderSchema>;

/** LLM Provider types (subset used for chat/evaluation) */
export const llmProviderSchema = z.enum(['anthropic', 'openai', 'gemini']);
export type LLMProvider = z.infer<typeof llmProviderSchema>;

/**
 * Schema for API key information (without exposing actual key).
 */
export const apiKeyInfoSchema = z.object({
  provider: z.string(),
  key_hint: z.string(),
  is_valid: z.boolean(),
  is_required: z.boolean().default(false),
});
export type ApiKeyInfo = z.infer<typeof apiKeyInfoSchema>;

/**
 * Schema for list of API keys response.
 */
export const apiKeyListResponseSchema = z.object({
  keys: z.array(apiKeyInfoSchema),
  openai_configured: z.boolean(),
});
export type ApiKeyListResponse = z.infer<typeof apiKeyListResponseSchema>;

/**
 * Schema for setting an API key request.
 */
export const setApiKeyRequestSchema = z.object({
  api_key: z.string().min(10, 'API key must be at least 10 characters'),
});
export type SetApiKeyRequest = z.infer<typeof setApiKeyRequestSchema>;

/**
 * Schema for set API key response.
 */
export const setApiKeyResponseSchema = z.object({
  provider: z.string(),
  key_hint: z.string(),
  is_valid: z.boolean(),
  message: z.string(),
});
export type SetApiKeyResponse = z.infer<typeof setApiKeyResponseSchema>;

/**
 * Schema for validating an API key request.
 */
export const validateKeyRequestSchema = z.object({
  provider: aiProviderSchema,
  api_key: z.string().min(10, 'API key must be at least 10 characters'),
});
export type ValidateKeyRequest = z.infer<typeof validateKeyRequestSchema>;

/**
 * Schema for validate key response.
 */
export const validateKeyResponseSchema = z.object({
  provider: z.string(),
  is_valid: z.boolean(),
  message: z.string(),
});
export type ValidateKeyResponse = z.infer<typeof validateKeyResponseSchema>;

// =============================================================================
// Agent Config Schemas
// =============================================================================

/**
 * Schema for agent configuration response.
 */
export const agentConfigResponseSchema = z.object({
  default_llm_provider: llmProviderSchema.nullable().default('anthropic'),
  default_llm_model: z.string().nullable().default(null),
  chat_provider: llmProviderSchema.nullable().default(null),
  chat_model: z.string().nullable().default(null),
  scorer_provider: llmProviderSchema.nullable().default(null),
  scorer_model: z.string().nullable().default(null),
  // Read-only (always OpenAI)
  embeddings_provider: z.string().default('openai'),
  embeddings_model: z.string().default('text-embedding-3-small'),
});
export type AgentConfigResponse = z.infer<typeof agentConfigResponseSchema>;

/**
 * Schema for updating agent configuration.
 */
export const updateAgentConfigRequestSchema = z.object({
  default_llm_provider: llmProviderSchema.nullable().optional(),
  default_llm_model: z.string().nullable().optional(),
  chat_provider: llmProviderSchema.nullable().optional(),
  chat_model: z.string().nullable().optional(),
  scorer_provider: llmProviderSchema.nullable().optional(),
  scorer_model: z.string().nullable().optional(),
});
export type UpdateAgentConfigRequest = z.infer<typeof updateAgentConfigRequestSchema>;

// =============================================================================
// Available Models Schemas
// =============================================================================

/**
 * Schema for a model option.
 */
export const modelOptionSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string(),
});
export type ModelOption = z.infer<typeof modelOptionSchema>;

/**
 * Schema for provider models.
 */
export const providerModelsSchema = z.object({
  provider: z.string(),
  provider_name: z.string(),
  models: z.array(modelOptionSchema),
});
export type ProviderModels = z.infer<typeof providerModelsSchema>;

/**
 * Schema for available models response.
 */
export const availableModelsResponseSchema = z.object({
  providers: z.array(providerModelsSchema),
});
export type AvailableModelsResponse = z.infer<typeof availableModelsResponseSchema>;

// =============================================================================
// Setup Status Schema
// =============================================================================

/**
 * Schema for setup status response.
 */
export const setupStatusResponseSchema = z.object({
  is_complete: z.boolean(),
  openai_configured: z.boolean(),
  llm_configured: z.boolean(),
  missing: z.array(z.string()),
});
export type SetupStatusResponse = z.infer<typeof setupStatusResponseSchema>;

File: backend/app/features/settings/settings_schemas.py
"""
User Settings Pydantic Schemas

Defines request/response models for API key and agent configuration endpoints.
"""

from typing import Optional, Literal
from pydantic import BaseModel, Field


# =============================================================================
# API KEY SCHEMAS
# =============================================================================

AIProviderType = Literal["openai", "anthropic", "gemini"]


class ApiKeyInfo(BaseModel):
    """API key information (without exposing the actual key).
    
    Attributes:
        provider: AI provider name.
        key_hint: Last 4 characters of the key for identification.
        is_valid: Whether the key has been validated.
        is_required: Whether this key is required (OpenAI for embeddings).
    """
    provider: str
    key_hint: str
    is_valid: bool
    is_required: bool = False


class ApiKeyListResponse(BaseModel):
    """Response listing all user API keys.
    
    Attributes:
        keys: List of API key info (hints only, not actual keys).
        openai_configured: Whether OpenAI API key is configured (required).
    """
    keys: list[ApiKeyInfo]
    openai_configured: bool


class SetApiKeyRequest(BaseModel):
    """Request to set/update an API key.
    
    Attributes:
        api_key: The API key to store (will be encrypted).
    """
    api_key: str = Field(..., min_length=10, description="API key to store")


class SetApiKeyResponse(BaseModel):
    """Response after setting an API key.
    
    Attributes:
        provider: AI provider name.
        key_hint: Last 4 characters of the stored key.
        is_valid: Whether the key was validated successfully.
        message: Status message.
    """
    provider: str
    key_hint: str
    is_valid: bool
    message: str


class ValidateKeyRequest(BaseModel):
    """Request to validate an API key without storing.
    
    Attributes:
        provider: AI provider to validate against.
        api_key: The API key to validate.
    """
    provider: AIProviderType
    api_key: str = Field(..., min_length=10)


class ValidateKeyResponse(BaseModel):
    """Response from API key validation.
    
    Attributes:
        provider: AI provider name.
        is_valid: Whether the key is valid.
        message: Validation result message.
    """
    provider: str
    is_valid: bool
    message: str


# =============================================================================
# AGENT CONFIG SCHEMAS
# =============================================================================

LLMProviderType = Literal["anthropic", "openai", "gemini"]


class AgentConfigResponse(BaseModel):
    """User's agent configuration preferences.
    
    Note: Embeddings are always OpenAI (not configurable by user).
    
    Attributes:
        default_llm_provider: User's preferred LLM provider.
        default_llm_model: User's preferred model for the provider.
        chat_provider: Provider for chat/RAG (overrides default).
        chat_model: Model for chat/RAG.
        scorer_provider: Provider for CV evaluation (overrides default).
        scorer_model: Model for CV evaluation.
    """
    default_llm_provider: Optional[LLMProviderType] = "anthropic"
    default_llm_model: Optional[str] = None
    
    # Per-agent overrides (optional)
    chat_provider: Optional[LLMProviderType] = None
    chat_model: Optional[str] = None
    scorer_provider: Optional[LLMProviderType] = None
    scorer_model: Optional[str] = None
    
    # Read-only (not user configurable)
    embeddings_provider: str = "openai"
    embeddings_model: str = "text-embedding-3-small"


class UpdateAgentConfigRequest(BaseModel):
    """Request to update agent configuration.
    
    Note: Users cannot change embeddings provider (always OpenAI).
    
    Attributes:
        default_llm_provider: Preferred LLM provider for all agents.
        default_llm_model: Preferred model name.
        chat_provider: Override for chat agent.
        chat_model: Override model for chat.
        scorer_provider: Override for scorer agent.
        scorer_model: Override model for scorer.
    """
    default_llm_provider: Optional[LLMProviderType] = None
    default_llm_model: Optional[str] = None
    chat_provider: Optional[LLMProviderType] = None
    chat_model: Optional[str] = None
    scorer_provider: Optional[LLMProviderType] = None
    scorer_model: Optional[str] = None


# =============================================================================
# AVAILABLE MODELS
# =============================================================================

class ModelOption(BaseModel):
    """A model option for a provider."""
    id: str
    name: str
    description: str


class ProviderModels(BaseModel):
    """Available models for a provider."""
    provider: str
    provider_name: str
    models: list[ModelOption]


class AvailableModelsResponse(BaseModel):
    """Response listing all available LLM providers and models."""
    providers: list[ProviderModels]
